### ISSUE TITLE
Reduction gets confused: A doesn't exist but it complains about B

### ISSUE THREAD

**Issue by blais** [beancount/beancount#256]
**[Original report](https://bitbucket.org/blais/beancount/issue/256) by Martin Michlmayr (Bitbucket: [tbm13](https://bitbucket.org/tbm13), GitHub: [tbm](https://github.com/tbm)).**

----------------------------------------

It took me a while to figure out what's going on because the error is misleading.  I have  2  DE0002635307 @ 36.67 EUR.  And there's 1 DE0002635307 @ 32.86 which isn't properly handled because I reduce it before it exists.  But the error I get is about the 2  DE0002635307 @ 36.67 EUR!

```
2000-01-01 open Assets:Investments
2000-01-01 open Equity:Opening-Balance
2000-01-01 open Assets:Transfer

2016-01-01 * "Opening balance"
  Assets:Transfer             2 DE0002635307 {36.67 EUR, 2007-09-18}
  Equity:Opening-Balance

2016-01-21 * "Transfer in"
  Assets:Investments          2 DE0002635307 {36.67 EUR, 2007-09-18}
  Assets:Transfer            -2 DE0002635307 {36.67 EUR, 2007-09-18}
  Assets:Investments          1 DE0002635307 {32.86 EUR, 2014-10-24}
  Assets:Transfer            -1 DE0002635307 {32.86 EUR, 2014-10-24}

2016-01-22 * "Transfer out"
  Assets:Transfer             2 DE0002635307 {36.67 EUR, 2007-09-18}
  Assets:Investments         -2 DE0002635307 {36.67 EUR, 2007-09-18}
```

Error:

```
/home/tbm/t1:15:      No position matches "Posting(account='Assets:Transfer', units=2
 DE0002635307, cost=CostSpec(number_per=Decimal('36.67'), number_total=None,
 currency='EUR', date=datetime.date(2007, 9, 18), label=None, merge=False), price=None, 
flag=None, meta={'lineno': 16, 'filename': '/home/tbm/t1'})" against balance (-1
 DE0002635307 {32.86 EUR, 2014-10-24})

   2016-01-22 * "Transfer out"
     Assets:Transfer      2 DE0002635307 {36.67 EUR, 2007-09-18}
     Assets:Investments  -2 DE0002635307 {36.67 EUR, 2007-09-18}
```

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

That's not what's happening here. A few preliminaries to make sure we're on the same page:
 
- For accounts with a negative balance a "reduction" is adding to them (e.g. if the balance is negative, you reduce by adding; in other words, "reduce" is always having to match against the existing lots of an inventory).
- The lot date (within {...}) doesn't count for ordering.

What's happening here is correct. Here's what's going on in the transfer account:

on 1/1 you're adding 2 DE {36.67 EUR}
on 1/21 you're removing them, they match, balance is empty
in the same txn on 1/21 you're removing a further 1 DE {32.86 EUR}. This time the balance is that lot.
on 1/22 you're reducing that by attempting to att 2 DE {36.67 EUR}. 

It can't find any; the only available lots are that -1 DE {32.86 EUR}.
Here's one way to debug this, see contents of transfer account after the second txn:


```
#!python


bean-doctor context test.beancount 11
Hash:1f5a2e9d9f28fc1e6b4f2f570b3c7535
Location: /tmp/test.beancount:9

------------ Balances before transaction

  Assets:Investments

  Assets:Transfer                2 DE0002635307 {36.67 EUR, 2007-09-18}


------------ Transaction

2016-01-21 * "Transfer in"
  Assets:Investments   2 DE0002635307 {36.67 EUR, 2007-09-18}  ;  73.34 EUR
  Assets:Transfer     -2 DE0002635307 {36.67 EUR, 2007-09-18}  ; -73.34 EUR
  Assets:Investments   1 DE0002635307 {32.86 EUR, 2014-10-24}  ;  32.86 EUR
  Assets:Transfer     -1 DE0002635307 {32.86 EUR, 2014-10-24}  ; -32.86 EUR



------------ Balances after transaction
```


* Assets:Investments             2 DE0002635307 {36.67 EUR, 2007-09-18}
* Assets:Investments             1 DE0002635307 {32.86 EUR, 2014-10-24}

* Assets:Transfer               -1 DE0002635307 {32.86 EUR, 2014-10-24}

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

WAI

**Comment by blais** [comment]
**Original comment by Martin Michlmayr (Bitbucket: [tbm13](https://bitbucket.org/tbm13), GitHub: [tbm](https://github.com/tbm)).**

----------------------------------------

Just for the record, Martin opened issue #261 to improve the error message.

Thanks for your detailed explanation.  I guess what you say makes sense but it's definitely very confusing for the user. It took me a while to figure out what's going on (the original transaction was more complicated).  So hopefully a future fix for as per #261 will make it clearer.
